[C10d] Support subgroups

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59111 [C10d] Support subgroups**
* #59693 Clang format distributed_test.py

Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test, and also clang format `distributed_c10d.py`.

Note that this API does not accept another overall main group. Like APEX API `create_syncbn_process_group` [here](https://nvidia.github.io/apex/_modules/apex/parallel.html), always uses the global world size and should only be applied when CUDA is available. 

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

Commit:
[C10d] Support subgroups

Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test.

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test.

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test.

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            should be given as a lowercase string (e.g., ``"gloo"``), which can
+            also be accessed via :class:`Backend` attributes (e.g.,
+            ``Backend.GLOO``). If ``None`` is passed in, the backend
+            corresponding to the default process group will be used. Default is
```
Should we instead default this to the backend for the input `group`, if possible (and then lastly fallback to the default group backend)?

SciPioneer replied:
Good point!

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+        pg_options (ProcessGroupOptions, optional): process group options
+            specifying what additional options need to be passed in during
+            the construction of specific process groups. i.e. for the ``nccl``
+            backend, is_high_priority_stream can be specified so that process
```
nit: backtick `high_priority_stream`

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            group can pick up high priority cuda streams.
+
+    Returns:
+        The subgroup containing the current rank, and all the subgroups used for cleanup.
```
Should we specify the return type? Seems like it is `Tuple[Group, List[Group]]`?

SciPioneer replied:
I prefer keeping the consistency at least in this PR. If we want to add more typing hint, we can do it another PR for all the functions.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+    """
+    subgroups = []
+    cur_subgroup = None
+    if ranks_per_subgroup_list:
```
Should we validate the input structure in some way? For example this code wouldn't execute if `ranks_per_subgroup_list=[]`, not sure what's the desired behavior there in terms of throwing an error or proceeding with the default behavior.

SciPioneer replied:
I added some description in the docstring to treat an empty list same as `None`.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+
+            rank = get_rank()
+            if rank in ranks:
+                cur_subgroup = subgroup
```
What if there is an overlap? It looks like we return the last subgroup that the rank is a part of, should we return all of them somehow? Or explicitly forbid overlaps?

SciPioneer replied:
Good question. Actually we can support overlapping subgroups. Here since we want to return a current subgroup, overlap is forbidden. Otherwise, there can be multiple current subgroups.

I don't see such a need at this time, so let's keep things simple.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+        else (os.cpu_count() or 1)
+    )
+    # If the given main group does not use all the devices,
+    # then the number of devices used by the subgroup should exceed the world size.
```
"should exceed the world size" --> "should _not_ exceed the world size"? At least that what it seems like from the next line, which takes num_devices_per_machine to be the smaller of the group's world size and num devices per machine.

SciPioneer replied:
Good catch!

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+        subgroups.append(subgroup)
+
+        rank = get_rank(group)
+        if rank >= start_rank and rank < end_rank:
```
nit: `rank in per_machine_ranks` could be a bit clearer.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+        return cur_subgroup, subgroups
+
+    # By default, create one subgroup per machine.
+    # Assume that each machine has the same number of devices.
```
Can we add this assumption to the docs? Also, what could break if this isn't true? It would be good to have an idea about what could happen to respond to bugs/user reports etc.

SciPioneer replied:
Added a warning.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+    num_devices_per_machine = (
+        torch.cuda.device_count()
+        if torch.cuda.is_available()
+        else (os.cpu_count() or 1)
```
I'm actually not sure about what the expected behavior should be for CPU-only use cases. I don't think I've seen a use case where one rank is spawned for each CPU core, which is what this code seems to do/ Is there such a use case? If not, should we always just use `1` if cuda is not available?

SciPioneer replied:
If a machine has multiple CPUs and no GPU, then each subgroup can have multiple ranks where each rank corresponds to a CPU. I think this can avoid surprising users and keep the subgroup creation behavior consistent in GPU and CPU-only scenarios.

I haven't seen any CPU-only use case yet, so we cannot easily know what will be expected behavior. I will prefer keeping the code this way.

Added a warning to declare that this API is experimental and subject to change, so we can make changes in the future.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+    )
+    # If the given main group does not use all the devices,
+    # then the number of devices used by the subgroup should exceed the world size.
+    num_devices_per_machine = min(get_world_size(group), num_devices_per_machine)
```
Is this logic correct? Say there are 8 GPUs per machine and 4 machines. If `group` uses 4 GPUs across each machine, then `num_devices_per_machine=8`, but should it be 4 instead? Or are we indeed okay with still using all GPUs on each machine even though `group` doesn't (in which case we shouldn't need this check)?

SciPioneer replied:
I decide to break this util function into two functions, one for equi-size partition, one for manually specifying subgroups.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            )
+            subgroups.append(subgroup)
+
+            rank = get_rank()
```
Does `group` need to be passed into all these get_rank calls, otherwise default pg will always be used?

SciPioneer replied:
Added. Thanks!

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            )
+            subgroups.append(subgroup)
+
+            rank = get_rank()
```
also, what happens if a rank given in `ranks` is not part of the given `group`? Should we throw an appropriate error? Can we also add a test for this condition?

SciPioneer replied:
Added a check and a test `test_new_subgroups_input_rank_exceeds_world_size_nccl`.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+        start_rank = machine_rank * num_devices_per_machine
+        end_rank = start_rank + num_devices_per_machine
+        per_machine_ranks = list(range(start_rank, end_rank))
+        subgroup = new_group(
```
if `ranks_per_subgroup_list=None`, then we can do the above computation to compute per machine ranks list, and then re-use the above logic that calls `new_group` and eliminate some of the code duplication here.

SciPioneer replied:
If `ranks_per_subgroup_list=None`, then the first if statement will not be executed, and per-machine subgroups will be created.

I don't see much code duplication in the implementations for these 2 cases. Which lines of code do you want to deduplicate?

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+    backend=None,
+    pg_options=None,
+):
+    """
```
Thanks for adding these docs! Can you also add a basic example into the docs on how the API can be used and an example of the cleanup process?

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -789,6 +789,36 @@ def test_backend_group(self):
+                if device_id == 0 or device_id == 2:
+                    self.assertEqual(cur_subgroup, subgroups[0])
+                else:
+                    self.assertEqual(cur_subgroup, subgroups[1])
```
Can we add tests that use the `_rank_not_in_group` API to verify that the appropriate ranks are in/not in the relevant groups?

SciPioneer replied:
Added

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -789,6 +789,36 @@ def test_backend_group(self):
+
+            # Divide into two subgroups, each of which has two processors.
+            # The device ids in the same subgroup are not neighboring numbers.
+            cur_subgroup, subgroups = dist.new_subgroups(ranks_per_subgroup_list=[[0, 2], [1, 3]], group=group)
```
Can we test non-default group as well? It looks like the type of `group` here is `list`, but I don't think the API expects this type for `group`. It might crash in the part of code that does the per-machine ranks logic?

SciPioneer replied:
No non-default group is accepted now.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -789,6 +789,36 @@ def test_backend_group(self):
+        @skip_if_lt_x_gpu(int(os.environ["WORLD_SIZE"]))
+        def test_new_default_subgroups_nccl(self):
+            world_size = dist.get_world_size()
+            cur_subgroup, subgroups = dist.new_subgroups()
```
Would be good to test passing in non-default group for this part of the test and below as well

SciPioneer replied:
Non-default group is not accepted now.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -789,6 +789,36 @@ def test_backend_group(self):
+            self.assertFalse(dist._rank_not_in_group(cur_subgroup))
+
+        @skip_if_lt_x_gpu(4)
+        def test_new_2subgroups_nccl(self):
```
Can we also add a tests for:
1) malformed input, i.e. bad `ranks_per_subgroup_list`
2) validate logic when there is overlapping (i.e. rank occurs more than once in the sublists)
3) What happens when we pass in a rank into `rank_per_subgroup_list` that is not part of the passed in `group`?

rohan-varma reviewed (commented):
Thanks for this great change! Just have a couple of comments/questions and suggestions on testing

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test, and also clang format `distributed_c10d.py`.

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test, and also clang format `distributed_c10d.py`.

Note that this API does not accept another overall main group. Like APEX API `create_syncbn_process_group` [here](https://nvidia.github.io/apex/_modules/apex/parallel.html), always uses the global world size and should only be applied when CUDA is available. 

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test, and also clang format `distributed_c10d.py`.

Note that this API does not accept another overall main group. Like APEX API `create_syncbn_process_group` [here](https://nvidia.github.io/apex/_modules/apex/parallel.html), always uses the global world size and should only be applied when CUDA is available. 

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test, and also clang format `distributed_c10d.py`.

Note that this API does not accept another overall main group. Like APEX API `create_syncbn_process_group` [here](https://nvidia.github.io/apex/_modules/apex/parallel.html), always uses the global world size and should only be applied when CUDA is available. 

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test, and also clang format `distributed_c10d.py`.

Note that this API does not accept another overall main group. Like APEX API `create_syncbn_process_group` [here](https://nvidia.github.io/apex/_modules/apex/parallel.html), always uses the global world size and should only be applied when CUDA is available. 

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2837,3 +2829,241 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+    pg_options=None,
+):
+    """
+    Creates GPU subgroups by divding the global world, where the division is specified by
```
nit: dividing (typo)

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2837,3 +2829,241 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+    Examples:
+        >>> # Create two subgroups, where each has 2 processes.
+        >>> cur_subgroup, subgroups = dist.new_subgroups(ranks=[[0, 2], [1, 3]])
+        >>> tensor = torch.ones(1) * dist.get_rank()
```
nit: `tensor = torch.ones(1, device=rank)` because this API requires CUDA

rohan-varma reviewed (approved):
Awesome, looks great! Does this PR need a ci-all to ensure the 4 gpu tests are passing, as I don't think we run those on PRs?

SciPioneer commented:
> Awesome, looks great! Does this PR need a ci-all to ensure the 4 gpu tests are passing, as I don't think we run those on PRs?

I've already ran the tests locally on an 8-GPU mahcine.

Commit:
Update on "[C10d] Support subgroups"


Create a util function for initializing subgroups. By default, each subgroup contains all the ranks within a machine. This util function can be used by both local SGD and SyncBatchNorm optimization.

Additionally, clang format `distributed/__init__.py` after importing `_rank_not_in_group` which is used by the unit test, and also clang format `distributed_c10d.py`.

Note that this API does not accept another overall main group. Like APEX API `create_syncbn_process_group` [here](https://nvidia.github.io/apex/_modules/apex/parallel.html), always uses the global world size and should only be applied when CUDA is available. 

#Closes: https://github.com/pytorch/pytorch/issues/53962

Differential Revision: [D28495672](https://our.internmc.facebook.com/intern/diff/D28495672/)

[ghstack-poisoned]
